File: Data Structures & Algorithms/palindrome-partitioning/submission-0.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> res =new ArrayList<>();
        rec(new ArrayList<>(), s, res );
        return res;
    }

    public void rec(List<String> found, String remaining, List<List<String>> results){
        if (remaining.isEmpty())
            results.add(found);
        
        for (int i = 1; i <= remaining.length(); i++){
            if (isPal(remaining.substring(0,i))){
                List<String> cur = new ArrayList<>(found);
                cur.add(remaining.substring(0,i));
                rec(cur,remaining.substring(i), results);
            }

        }

    }

    public boolean isPal(String s){
        int left = 0, right = s.length()-1;

        //System.out.printf("testing %s\n",s);
        while (left <= s.length() /2){
            if (s.charAt(left) != s.charAt(right))
                return false;
            left++;
            right--;
        }
        
        return true;
    }
}

File: Data Structures & Algorithms/decode-ways/submission-0.go
func numDecodings(s string) int {
    cache := map[int]int{len(s):1}

	return rec(s, 0, cache)
}

func rec(s string, i int, cache map[int]int) int{
	if c, ok := cache[i]; ok {
		return c;
	}

	if s[i] == '0' {
        return 0
    }

	res := rec(s, i + 1, cache)
	if i + 1 < len(s) {
		if s[i] == '1' || (s[i] == '2' && s[i+1] < '7') {
			res += rec(s, i + 2, cache)
		}
	}
	cache[i] = res
	return cache[i]
}


